client/client_update_test.go

```diff
@@ -90,8 +101,326 @@ func testUpdateWithLocalCacheRemoteMissingMetadata(t *testing.T, forWrite bool)
+				require.True(t, bytes.Equal(expected, actual),
+					"%s: expected to update since only %s was unwritable", r, role)
+			}
+		}
+
+		os.RemoveAll(repo.baseDir)
+	}
+}
+
+// Update can succeed even if we cannot write any metadata to the repo (assuming
+// existing data in the repo)
+func TestUpdateSucceedsEvenIfCannotWriteExistingRepo(t *testing.T) {
+	s, err := testutils.NewMetadataSwizzler("docker.com/notary")
+	require.NoError(t, err)
+
+	ts := readOnlyServer(t, s.MetadataCache, s.Roles)
+	defer ts.Close()
+
+	// download existing metadata
+	repo := newBlankRepo(t, ts.URL)
+	defer os.RemoveAll(repo.baseDir)
+	_, err = repo.Update(false)
+	require.NoError(t, err)
+
+	origFileStore := repo.fileStore
+	for _, role := range s.Roles {
+		for _, forWrite := range []bool{true, false} {
+			// bump versions of everything on the server, to force everything to update
+			for _, r := range s.Roles {
+				require.NoError(t, s.OffsetMetadataVersion(r, 1))
+			}
+			require.NoError(t, s.UpdateSnapshotHashes())
+			require.NoError(t, s.UpdateTimestampHash())
+
+			// update fileStore
+			repo.fileStore = &unwritableStore{MetadataStore: origFileStore, roleToNotWrite: role}
+			_, err := repo.Update(forWrite)
+
+			if role == data.CanonicalRootRole {
+				require.Error(t, err) // because checkRoot loads root from cache to check hashes
+				continue
+			}
+			require.NoError(t, err)
+
+			for _, r := range s.Roles {
+				expected, err := s.MetadataCache.GetMeta(r, maxSize)
+				require.NoError(t, err, "problem getting expected metadata for %s", r)
+				actual, err := repo.fileStore.GetMeta(r, maxSize)
+				require.NoError(t, err, "problem getting repo metadata for %s", r)
+				if role == r {
+					require.False(t, bytes.Equal(expected, actual),
+						"%s: expected to not update because %s was unwritable", r, role)
+				} else {
+					require.True(t, bytes.Equal(expected, actual),
+						"%s: expected to update since only %s was unwritable", r, role)
+				}
+			}
+		}
+	}
+}
+
+type messUpMetadata func(role string) error
+
+// If a repo has corrupt metadata (in that the hash doesn't match the snapshot) or
+// missing metadata, an update will replace all of it
+func TestUpdateReplacesCorruptOrMissingMetadata(t *testing.T) {
+	s, err := testutils.NewMetadataSwizzler("docker.com/notary")
+	require.NoError(t, err)
+
+	ts := readOnlyServer(t, s.MetadataCache, s.Roles)
+	defer ts.Close()
+
+	repo := newBlankRepo(t, ts.URL)
+	defer os.RemoveAll(repo.baseDir)
+
+	_, err = repo.Update(false) // ensure we have all metadata to start with
+	require.NoError(t, err)
+
+	// we want to swizzle the local cache, not the server, so create a new one
+	swizzler := testutils.MetadataSwizzler{
+		MetadataCache: repo.fileStore,
+		CryptoService: repo.CryptoService,
+		Roles:         s.Roles,
+	}
+
+	waysToMessUp := map[string]messUpMetadata{
+		"invalid JSON":     swizzler.SetInvalidJSON,
+		"missing metadata": swizzler.DeleteMetadata,
+	}
+	for _, role := range s.Roles {
```

Comment:
haven't looked into this deeply, but is it possible that it would take a lot of time since there are 4 nested `for`